### PR TITLE
Unified many implementations of common interfaces into abstract classes

### PR DESCRIPTION
**Pull request by yairm210**
Here it is ladles and jellyspoons, the abstract classes that probably should have been put in place ages ago.
Not only does this decrease the number of repetitive lines in the codebase and unify existing implementations, it will also allow us future, ruleset-object specific implementations.

### CODE REVIEW

File: core/src/com/unciv/models/ruleset/RulesetObject.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+// Same, but inherits from NamedStats - I couldn't find a way to unify the declarations but this is fine
+abstract class RulesetStatsObject: NamedStats(), IHasUniques, ICivilopediaText {
+    override var uniques = ArrayList<String>() // Can not be a hashset as that would remove doubles
+    @delegate:Transient
+    override val uniqueObjects: List<Unique> by lazy {
+        uniques.map { Unique(it, getUniqueTarget(), name) }
+    }
+    override var civilopediaText = listOf<FormattedLine>()
+}
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/models/ruleset/RulesetObject.kt:28]: If prefer github to not display that red traffic sign when a file doesn't end in a linefeed (though the [kotlinlang.org conventions](https://kotlinlang.org/docs/coding-conventions.html#formatting) do not say anything on that matter) :laughing:

**Review by SomeTroglodyte — CHANGES_REQUESTED** [review]
:+1:  - one "food for thought": For I while I had both `ICivilopediaText` and `CivilopediaText`, interface and abstract with the default implementations for fields. I dropped that mainly because you can implement as many ifaces as you want, but inherit only one superclass, and didn't want to impose limitations.

Here, it's definitely good this now provides *the* base, but maybe having the interface too as clearer documentation / for use in parameter passing...

Wait a minute - RulesetObject and RulesetStatsObject are independent! Ahem, changing Approve to Request changes...

`interface IRulesetObject: INamed, IHasUniques, ICivilopediaText` does compile, at a minimum...

**Comment by yairm210** [comment]
How would you suggest I unify them? Just with an interface?

**Review by SomeTroglodyte — APPROVED** [review]
Exactly. Having ***one*** type - whatever kind of - apply to ***all*** of them is the goal, right? I can talk haughtily like that easily, as I won't be doing any testing on Unicv for a few hours :wink:

**Pull request merged**